src/edu/washington/escience/myria/operator/SamplingDistribution.java

```diff
@@ -0,0 +1,328 @@
+   * @param sampleSize
+   *          total samples to create a distribution for.
+   * @param sampleType
+   *          the type of sampling distribution to create
+   * @param child
+   *          extracts (WorkerID, PartitionSize, StreamSize) information from
+   *          this child.
+   * @param randomSeed
+   *          value to seed the random generator with. null if no specified seed
+   */
+  public SamplingDistribution(Operator child, int sampleSize,
+      SamplingType sampleType, Long randomSeed) {
+    this(child, sampleType, randomSeed);
+    Preconditions.checkArgument(sampleSize >= 0,
+        "Sample Size must be >= 0: %s", sampleSize);
+    this.sampleSize = sampleSize;
+  }
+
+  /**
+   * Instantiate a SamplingDistribution operator using a percentage of total
+   * tuples.
+   *
+   * @param samplePercentage
+   *          percentage of total samples to create a distribution for.
+   * @param sampleType
+   *          the type of sampling distribution to create
+   * @param child
+   *          extracts (WorkerID, PartitionSize, StreamSize) information from
+   *          this child.
+   * @param randomSeed
+   *          value to seed the random generator with. null if no specified seed
+   */
+  public SamplingDistribution(Operator child, float samplePercentage,
+      SamplingType sampleType, Long randomSeed) {
+    this(child, sampleType, randomSeed);
+    this.isPercentageSample = true;
+    this.samplePercentage = samplePercentage;
+    Preconditions.checkArgument(samplePercentage >= 0
+        && samplePercentage <= 100,
+        "Sample Percentage must be >= 0 && <= 100: %s", samplePercentage);
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws DbException {
+    // Drain out all the worker info.
+    while (!getChild().eos()) {
+      TupleBatch tb = getChild().nextReady();
+      if (tb == null) {
+        if (getChild().eos()) {
+          break;
+        }
+        return null;
+      }
+      extractWorkerInfo(tb);
+    }
+
+    // Convert samplePct to sampleSize if using a percentage sample.
+    if (isPercentageSample) {
+      sampleSize = Math.round(totalTupleCount * (samplePercentage / 100));
+    }
+    Preconditions.checkState(sampleSize >= 0 && sampleSize <= totalTupleCount,
+        "Cannot extract %s samples from a population of size %s", sampleSize,
+        totalTupleCount);
+
+    // Generate a sampling distribution across the workers.
+    if (sampleType == SamplingType.WithReplacement) {
+      withReplacementDistribution(workerInfo, totalTupleCount, sampleSize);
+    } else if (sampleType == SamplingType.WithoutReplacement) {
+      withoutReplacementDistribution(workerInfo, totalTupleCount, sampleSize);
+    } else {
+      throw new DbException("Invalid sampleType: " + sampleType);
+    }
+
+    // Build and return a TupleBatch with the distribution.
+    // Assumes that the sampling information can fit into one tuple batch.
+    IntColumnBuilder wIDs = new IntColumnBuilder();
```

Comment:
You could also use `ColumnFactory.allocateColumns(SCHEMA)` to create the list of `ColumnBuilder`s in one line, although you might sacrifice some explicitness unless you named the column indices.